src/edu/csus/ecs/pc2/clics/API202306/CLICSContestAccess.java

```diff
@@ -0,0 +1,167 @@
+                if(Utilities.isDebugMode()) {
+                    System.err.println("No endpoint access for service " + service);
+                }
+                controller.getLog().log(Log.WARNING, "No endpoint access for service " + service);
+            }
+        }
+        if(!epList.isEmpty()) {
+            endpoints = epList.toArray(new CLICSEndpoint[0]);
+        }
+    }
+
+    /**
+     * Create a fully qualified class name for the supplied web service endpoint.
+     * eg. Group would create: edu.csus.ecs.pc2clics.API202306.GroupService
+     * Then call its getEndpointProperties(SecurityContext) static method to see what properties the
+     * service supports for the supplied SecurityContext (user).
+     * Note: If Java reflection makes you queasy, please stop here and just skip reading the code in this method.
+     *
+     * @param epList List to add the CLICSEndpointObject to, if one was returned
+     * @param serviceName The service name only (excluding the trailing "Service")
+     * @param sc User's security context
+     * @return true if this endpoint is accessible by the user in the SecurityContext
+     */
+    private boolean invokeServiceEndpointAccessMethod(ArrayList<CLICSEndpoint> epList, String serviceName, SecurityContext sc) {
+
+        Class c = this.getClass();
+        String className = c.getPackage().getName() + "." + serviceName + "Service";
+        try {
+            Class<?> newClass = Class.forName(className);
+            try {
+                Method m = newClass.getMethod(ENDPOINT_ACCESS_METHOD, SecurityContext.class);
+                if(m != null) {
+                    // invoke static method with the SecurityContext as an argument
+                    Object epRet = m.invoke(null, sc);
+                    if(epRet instanceof CLICSEndpoint) {
+                        // we got one back, so add it to the list
+                        CLICSEndpoint ep = (CLICSEndpoint)epRet;
+                        epList.add(ep);
+                        return(true);
+                    } else {
+                        controller.getLog().log(Log.WARNING, "CLICS Service class " + className + "." + ENDPOINT_ACCESS_METHOD + "(SecurityContext) does not return a CLICSEndpoint");
+                    }
+                }
+            } catch (Exception eMethod) {
+                controller.getLog().log(Log.WARNING, "Unable to invoke method " + className + "." + ENDPOINT_ACCESS_METHOD + "(SecurityContext)");
+            }
+        } catch (Exception e) {
+            controller.getLog().log(Log.WARNING, "Unable to load CLICS Service class " + className);
+        }
+
+        return false;
+    }
+
+    public String toJSON() {
+
+        try {
+            ObjectMapper mapper = JSONUtilities.getObjectMapper();
+            return mapper.writeValueAsString(this);
+        } catch (Exception e) {
+            return "Error creating JSON for contest state info " + e.getMessage();
```

Comment:
Corresponding to an earlier comment, I'm not sure why this message string refers to "contest state"... It seems to me that "Contest Access" is not the same thing as "Contest State" (and this class is about _Access_).

Reply:
Yes, copy paste error.  Fixed in next push.